File: algorithm-training/typescript/leet/binarySearch/FirstBadVersion.spec.ts
import { firstBad } from './FirstBadVersion';

describe('firstBad', () => {
    it('should return the first bad version', () => {
        const isBadVersion = (version: number) => version >= 4;
        const n = 5;
        expect(firstBad(isBadVersion)(n)).toBe(4);
    });

    it('should return the first bad version', () => {
        const isBadVersion = (version: number) => version >= 1;
        const n = 1;
        expect(firstBad(isBadVersion)(n)).toBe(1);
    });

    it('should return the first bad version', () => {
        const isBadVersion = (version: number) => version >= 1;
        const n = 2;
        expect(firstBad(isBadVersion)(n)).toBe(1);
    });
});

File: algorithm-training/kotlin/src/commonTest/kotlin/leet/BinarySearch/FirstBadVersion.spec.kt
package leet.BinarySearch

import kotlin.test.assertEquals
import kotlin.test.Test

class FirstBadVersionTest {

    @Test
    fun testFirstBadVersion() {
        // Mock the isBadVersion function
        val isBadVersion: (Int) -> Boolean = { version ->
            version >= 4 // Example: the first bad version is 4
        }

        val solution = FirstBadVersion()
        val n = 5 // Example: total versions

        // Verify that the first bad version is correctly identified
        val result = solution.firstBadVersion(n, isBadVersion)
        assertEquals(4, result)
    }
}